### PR TITLE
Integration of H5 family

### PR DESCRIPTION
**Pull request by Hish15**
Progress #318

### CODE REVIEW

File: cmake/stm32/h5.cmake [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    32K 640K 640K 640K
+)
+set(STM32_H5_CCRAM_SIZES 
+    0K  0K  0K  0K
+)
+
+stm32_util_create_family_targets(H5)
+
+target_compile_options(STM32::H5 INTERFACE 
+    -mcpu=cortex-m7 -mfloat-abi=hard -mfpu=fpv5-d16 -mthumb
```
Comment (PhilippHaefele) [inline cmake/stm32/h5.cmake:18]: Just a notice from someone following the project:  
As the H5 is a Cortex-M55 i did expect `-mcpu=cortex-m33` here and below.

Comment (atsju) [reply cmake/stm32/h5.cmake:18]: Correct. You mean M33 not M55 I suppose ?

File: docs/MAINTAIN.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Supporting new families
+
+## Write device file
+
+a file <family>.cmake must be added to the folder cmake/stm32
+This file containes the differents devices and the regex used to parse them.
+It also give information on the RAM and CCRAM available
+
```
Comment (atsju) [inline docs/MAINTAIN.md:8]:
```suggestion
Maintainer must check than Flash sizes correspond to usual suffix. Some families have exceptions.

```
+ add the suffix or hint file containing the sufix to size code

File: docs/MAINTAIN.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+## Add family to list of know families
+
+Update the list `STM32_SUPPORTED_FAMILIES_LONG_NAME` located in the cmake/stm32/common.cmake file
+
+
+## Add family to CI
+
+Update the cmake.yml workflow file to include the new family
```
Comment (atsju) [inline docs/MAINTAIN.md:16]: I'm not sure we already automated. Maintainer must add the list of HAL and LL somewhere

File: cmake/stm32/h5.cmake (hunk 1 above)
Comment (PhilippHaefele) [reply cmake/stm32/h5.cmake:18]: Yes, did correct it in the initial comment

Comment (Hish15) [reply cmake/stm32/h5.cmake]: Yes of course

File: docs/MAINTAIN.md (hunk 3 above)
Comment (Hish15) [reply docs/MAINTAIN.md:16]: HE have some basics tests yes. It does not pass for now, since I just started making the changes

Comment (atsju) [reply docs/MAINTAIN.md:16]: finHAL has been implemented. Please ignore this comment

**Comment by tcbennun** [comment]
What else needs to be done here?

**Comment by koniarik** [comment]
So, in the meantime I've added my attempt at this to my local  fork:

https://github.com/ObKo/stm32-cmake/compare/master...koniarik:stm32-cmake:master

What has to be done here to merge this? or should I pake PR with my fork?

File: .github/workflows/cmake.yml [hunk 4]
```diff
@@ -11,7 +11,7 @@ jobs:
     runs-on: ubuntu-20.04
     strategy:
       matrix:
-        family: [F0, F1, F2, F3, F4, F7, G0, G4, H7, L0, L1, L4, L5, U5, WB, WL, MP1]
+        family: [H5]
```
Comment (atsju) [inline .github/workflows/cmake.yml:14]: this is incorrect. H5 need to be added without removing others. I suppose it was for tests

**Review by atsju — COMMENTED** [review]
@Hish15  what is missing here from your POV ?

**Comment by Hish15** [comment]
Sorry guys, I will do the fixes from the PR and then it's good to go.

File: docs/MAINTAIN.md (hunk 2 above)
Comment (Hish15) [reply docs/MAINTAIN.md:8]: Can you make a suggestion with all the changes you want here ?

File: cmake/stm32/utilities.cmake [hunk 5]
```diff
@@ -43,10 +43,10 @@ endfunction()
...
-set(STM32_FETCH_CUBE_VERSIONS  v1.1.0 v1.11.2 v1.8.4  v1.9.3  v1.11.2 v1.26.1  v1.16.1 v1.4.1  v1.4.0  v1.9.0   v1.12.0 v1.10.3 v1.17.0 v1.4.0 1.5.0  v1.3.0 v1.12.0 v1.1.0)
-set(STM32_FETCH_CMSIS_VERSIONS v1.1.0 v2.3.5  v4.3.3  v2.2.5  v2.3.5  v2.6.6   v1.2.6  v1.4.0  v1.2.1  v1.10.0  v1.9.1  v2.3.2  v1.7.1  v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
-set(STM32_FETCH_HAL_VERSIONS   v1.1.0 v1.7.5  v1.1.8  v1.2.7  v1.5.5  v1.7.12  v1.2.9  v1.4.1  v1.2.1  v1.10.0  v1.10.4 v1.4.4  v1.13.0 v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
+set(STM32_FETCH_FAMILIES       C0     F0      F1      F2      F3      F4       F7      G0      G4      H5      H7       L0      L1      L4      L5     MP1    U5     WB      WL    )
+set(STM32_FETCH_CUBE_VERSIONS  v1.1.0 v1.11.2 v1.8.4  v1.9.3  v1.11.2 v1.26.1  v1.16.1 v1.4.1  v1.4.0  v1.0.0  v1.9.0   v1.12.0 v1.10.3 v1.17.0 v1.4.0 1.5.0  v1.3.0 v1.12.0 v1.1.0)
+set(STM32_FETCH_CMSIS_VERSIONS v1.1.0 v2.3.5  v4.3.3  v2.2.5  v2.3.5  v2.6.6   v1.2.6  v1.4.0  v1.2.1  v5.6.0  v1.10.0  v1.9.1  v2.3.2  v1.7.1  v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
```
Comment (Hish15) [inline cmake/stm32/utilities.cmake:48]:
```suggestion
set(STM32_FETCH_CMSIS_VERSIONS v1.1.0 v2.3.5  v4.3.3  v2.2.5  v2.3.5  v2.6.6   v1.2.6  v1.4.0  v1.2.1  v5.9.0  v1.10.0  v1.9.1  v2.3.2  v1.7.1  v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)

```

File: cmake/stm32/utilities.cmake [hunk 6]
```diff
@@ -43,10 +43,11 @@ endfunction()
...
-set(STM32_FETCH_CMSIS_VERSIONS v1.1.0 v2.3.5  v4.3.3  v2.2.5  v2.3.5  v2.6.6   v1.2.6  v1.4.0  v1.2.1  v1.10.0  v1.9.1  v2.3.2  v1.7.1  v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
-set(STM32_FETCH_HAL_VERSIONS   v1.1.0 v1.7.5  v1.1.8  v1.2.7  v1.5.5  v1.7.12  v1.2.9  v1.4.1  v1.2.1  v1.10.0  v1.10.4 v1.4.4  v1.13.0 v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
+set(STM32_FETCH_FAMILIES       C0     F0      F1      F2      F3      F4       F7      G0      G4      H5      H7       L0      L1      L4      L5     MP1    U5     WB      WL    )
+set(STM32_FETCH_CUBE_VERSIONS  v1.1.0 v1.11.2 v1.8.4  v1.9.3  v1.11.2 v1.26.1  v1.16.1 v1.4.1  v1.4.0  v1.0.0  v1.9.0   v1.12.0 v1.10.3 v1.17.0 v1.4.0 1.5.0  v1.3.0 v1.12.0 v1.1.0)
+set(STM32_FETCH_CMSIS_VERSIONS v1.1.0 v2.3.5  v4.3.3  v2.2.5  v2.3.5  v2.6.6   v1.2.6  v1.4.0  v1.2.1  v5.9.0  v1.10.0  v1.9.1  v2.3.2  v1.7.1  v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
+
+set(STM32_FETCH_HAL_VERSIONS   v1.1.0 v1.7.5  v1.1.8  v1.2.7  v1.5.5  v1.7.12  v1.2.9  v1.4.1  v1.2.1  v1.0.0  v1.10.0  v1.10.4 v1.4.4  v1.13.0 v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)
```
Comment (koniarik) [inline cmake/stm32/utilities.cmake:50]: Are we sure about the H5 versions?

CMSIS for H5 does not have 5.9 https://github.com/STMicroelectronics/cmsis_device_h5/tags
And the latest version for all repos based on my repo is v1.1.0

File: cmake/stm32/h5.cmake [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+stm32_util_create_family_targets(H5)
+
+target_compile_options(STM32::H5 INTERFACE 
+    -mcpu=cortex-m33 -mfloat-abi=hard -mfpu=fpv5-d16 -mthumb
+)
+
+target_link_options(STM32::H5 INTERFACE 
+    -mcpu=cortex-m33 -mfloat-abi=hard -mfpu=fpv5-d16 -mthumb
```
Comment (koniarik) [inline cmake/stm32/h5.cmake:22]: What is the source information for using fpv5-d16 FPU?

I was kinda not sure how to find this out so I opted to let STM32CubeMx generate an empty makefile project for STM32H5 and checked it's compiler options. CubeMx used fpv4-sp-d16 there.

**Review by koniarik — COMMENTED** [review]
I would like just to post few comments here to make sure this is perfectly valid, if you wnat any help/effort from my side feel free to ask for it. I am interested in making sure this passes :)

File: cmake/stm32/utilities.cmake (hunk 6 above)
Comment (Hish15) [reply cmake/stm32/utilities.cmake:50]: I confirm that 5.9 does not work. After further inspection this version is searched at this location : 
https://github.com/STMicroelectronics/cmsis_device_<FAMILY>

I made the mistake of reading the release note of STM32H5 repo. It states that the cmsis version is 5.9.0. 
https://github.com/STMicroelectronics/STM32CubeH5/blob/66f1a02802658fd01a404e770252b5ffc63986dd/Release_Notes.html#L116

Will trying to add H5 support I'm also documenting the process so it's easier next time. Thanks for your input

File: docs/MAINTAIN.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+a file <family>.cmake must be added to the folder cmake/stm32
+This file containes the differents devices and the regex used to parse them.
+It also give information on the RAM and CCRAM available
+
+## Add family to list of know families
+
+Update the list `STM32_SUPPORTED_FAMILIES_LONG_NAME` located in the cmake/stm32/common.cmake file
+
+
```
Comment (Hish15) [inline docs/MAINTAIN.md:13]: ## Update cube, cmsis and hal version in utilities.cmake

The versions to use can be found as follow:
 - Cube version : a valid tag from the repo https://github.com/STMicroelectronics/STM32Cube${FAMILY}
 - Cmsis version : a valid tag from the repo https://github.com/STMicroelectronics/cmsis_device_${FAMILY_L}
 - Hal version : a valid tag from the repo https://github.com/STMicroelectronics/stm32${FAMILY_L}xx_hal_driver

File: cmake/stm32/utilities.cmake (hunk 6 above)
Comment (Hish15) [inline cmake/stm32/utilities.cmake:50]:
```suggestion
set(STM32_FETCH_CMSIS_VERSIONS v1.1.0 v2.3.5  v4.3.3  v2.2.5  v2.3.5  v2.6.6   v1.2.6  v1.4.0  v1.2.1  v1.1.0  v1.10.0  v1.9.1  v2.3.2  v1.7.1  v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)

set(STM32_FETCH_HAL_VERSIONS   v1.1.0 v1.7.5  v1.1.8  v1.2.7  v1.5.5  v1.7.12  v1.2.9  v1.4.1  v1.2.1  v1.1.0  v1.10.0  v1.10.4 v1.4.4  v1.13.0 v1.0.4 cube  v1.3.0 v1.9.0  v1.1.0)

```

File: cmake/stm32/h5.cmake (hunk 7 above)
Comment (Hish15) [reply cmake/stm32/h5.cmake:22]: I could not find the answer easily. For what I can tell https://www.st.com/resource/en/programming_manual/pm0264-stm32-cortexm33-mcus-programming-manual-stmicroelectronics.pdf states that m33 devices support fpv5 instructions...
Not quite sure for this one... @atsju any input ?

File: docs/MAINTAIN.md (hunk 8 above)
Comment (Hish15) [inline docs/MAINTAIN.md:13]:
```suggestion
## Update cube, cmsis and hal version in utilities.cmake

The versions to use can be found as follow:
 - Cube version : a valid tag from the repo https://github.com/STMicroelectronics/STM32Cube${FAMILY}
 - Cmsis version : a valid tag from the repo https://github.com/STMicroelectronics/cmsis_device_${FAMILY_L}
 - Hal version : a valid tag from the repo https://github.com/STMicroelectronics/stm32${FAMILY_L}xx_hal_driver
 
```

File: cmake/stm32/utilities.cmake (hunk 6 above)
Comment (koniarik) [reply cmake/stm32/utilities.cmake]: All is fixed, nice!

File: cmake/stm32/h5.cmake (hunk 7 above)
Comment (atsju) [reply cmake/stm32/h5.cmake:22]: Sorry for delay. I don't have much time right now so I just answer the question here.

I have read this and others: https://community.st.com/t5/stm32cubemx-mcus/bug-in-cubemx-generating-makefile-missing-some-cpu-info/m-p/110239

For me correct answer is `fpv5-sp-d16`. v5 as pointed out by @hish15 but sp only as cubeMx and datasheets stating it's single point FPU. Do not hesitate to redo a test with latest cubeMx and to ask ST

**Review by atsju — APPROVED** [review]

**Comment by koniarik** [comment]
Just wanted to note: thanks for the MR and merge of this!

**Pull request merged**